src/extension_commands.ml

```diff
@@ -950,6 +950,185 @@ module Search_by_type = struct
+
+  let jump_to_range range text_editor =
+    let open Promise.Syntax in
+    let+ _ =
+      Window.showTextDocument
+        ~document:(TextEditor.document text_editor)
+        ~preserveFocus:true
+        ()
+    in
+    let selection =
+      let anchor = Range.start range in
+      let active = Range.end_ range in
+      Selection.makePositions ~anchor ~active
+    in
+    show_selection selection text_editor
+  ;;
+
+  module QuickPickItemWithRange = struct
+    type t =
+      { item : QuickPickItem.t
+      ; range : Range.t
+      }
+
+    let t_of_js js =
+      let range = Ojs.get_prop_ascii js "pl_range" |> Range.t_of_js in
+      let item = QuickPickItem.t_of_js js in
+      { item; range }
+    ;;
+
+    let t_to_js t =
+      let item = QuickPickItem.t_to_js t.item in
+      Ojs.set_prop_ascii item "pl_range" (Range.t_to_js t.range);
+      item
+    ;;
+  end
+
+  module QuickPick = Vscode.QuickPick.Make (QuickPickItemWithRange)
+
+  let display_results (results : Range.t list) text_editor client instance =
+    let open Promise.Syntax in
+    let selected_item = ref false in
+    let text_document = TextEditor.document text_editor in
+    let quickPickItems =
+      List.map results ~f:(fun range ->
+        let line = Position.line @@ Range.end_ range in
+        let item =
+          QuickPickItem.create
+            ~label:(Printf.sprintf "Line %d" line)
+            ~description:(TextLine.text @@ TextDocument.lineAt ~line text_document)
+            ()
+        in
+        { QuickPickItemWithRange.item; range })
+    in
+    let quickPick =
+      QuickPick.set
+        (Window.createQuickPick (module QuickPickItemWithRange) ())
+        ~title:"Typed Holes"
+        ~activeItems:[]
+        ~busy:false
+        ~enabled:true
+        ~placeholder:"Use arrow keys to preview / Select to jump to it"
+        ~selectedItems:[]
+        ~ignoreFocusOut:true
+        ~items:quickPickItems
+        ~buttons:[]
+        ()
+    in
+    let _disposable =
+      QuickPick.onDidChangeActive
+        quickPick
+        ~listener:(function
+          | { range; _ } :: _ ->
+            show_selection
+              (Selection.makePositions
+                 ~anchor:(Range.start range)
+                 ~active:(Range.end_ range))
+              text_editor
+          | _ -> ())
+        ()
+    in
+    let _disposable =
+      QuickPick.onDidAccept
+        quickPick
+        ~listener:(fun () ->
+          match QuickPick.selectedItems quickPick with
+          | Some (item :: _) ->
+            ignore
+              (let range = item.range in
+               let* () = jump_to_range range text_editor in
+               selected_item := true;
+               match Settings.(get server_typedHolesConstructAfterNavigate_setting) with
+               | Some true ->
+                 Construct.process_construct
+                   (Range.end_ range)
+                   text_editor
+                   client
+                   instance
```

Comment:
Testing this feature highlighted an issue with `process_construct` when there is no result, instead of showing a message it displays an empty quick pick which is very confusing. We should improve that behaviour in a separate PR :-)